src/lib.rs

```diff
@@ -451,136 +488,165 @@ impl Config {
+                                  .unwrap_or(String::new());
+                if features.contains("crt-static") {
+                    cmd.args.push("/MT".into());
+                } else {
+                    cmd.args.push("/MD".into());
+                }
+                match &opt_level[..] {
+                    "z" | "s" => cmd.args.push("/Os".into()),
+                    "2" => cmd.args.push("/O2".into()),
+                    // FIXME: -O1 does not exactly match the meaning of /O1 in MSVC. Rather than
+                    // being a flag for “generate sufficiently fast code but also compile fast”,
+                    // /O1 is instead “generate the smallest code in majority of cases”.
+                    "1" => cmd.args.push("/O1".into()),
+                    // FIXME: this behaves badly if opt_level is set to 3 (which is an option on
+                    // Gnu and Clang-family tools) or more. Consider using /Ox or /O2 instead.
+                    _ => {}
+                }
             }
-            if target.contains("i586") {
-                cmd.args.push("/ARCH:IA32".into());
+            ToolFamily::Gnu |
+            ToolFamily::Clang => {
+                cmd.args.push(format!("-O{}", opt_level).into());
+                if !nvcc {
+                    cmd.args.push("-ffunction-sections".into());
+                    cmd.args.push("-fdata-sections".into());
+                    if self.pic.unwrap_or(!target.contains("i686") &&
+                                          !target.contains("windows-gnu")) {
+                        cmd.args.push("-fPIC".into());
+                    }
+                } else if self.pic.unwrap_or(false) {
+                    cmd.args.push("-Xcompiler".into());
+                    cmd.args.push("\'-fPIC\'".into());
+                }
             }
-        } else if nvcc {
-            cmd.args.push(format!("-O{}", opt_level).into());
-        } else {
-            cmd.args.push(format!("-O{}", opt_level).into());
-            cmd.args.push("-ffunction-sections".into());
-            cmd.args.push("-fdata-sections".into());
         }
+
         for arg in self.envflags(if self.cpp {"CXXFLAGS"} else {"CFLAGS"}) {
             cmd.args.push(arg.into());
         }
 
-        if debug {
-            cmd.args.push(if msvc {"/Z7"} else {"-g"}.into());
+        if self.get_debug() {
+            cmd.args.push(cmd.family.debug_flag().into());
         }
 
-        if target.contains("-ios") {
-            self.ios_flags(&mut cmd);
-        } else if !msvc {
-            if target.contains("i686") || target.contains("i586") {
-                cmd.args.push("-m32".into());
-            } else if target.contains("x86_64") || target.contains("powerpc64") {
-                cmd.args.push("-m64".into());
+        // Target flags
+        match cmd.family {
+            ToolFamily::Clang => {
+                cmd.args.push(format!("--target={}", target).into());
```

Comment:
I think we differ from Clang in the specifics for some targets, right? Something like NEON on armv7 or along those lines? (I forget the details here)

I like the idea though of just communicating via the target...